### PR TITLE
Be able to configure conditional branches

### PR DESCRIPTION
**Pull request by kemper**
resolves manheim/terraform-pipeline#172

### CODE REVIEW

**Comment by kemper** [comment]
I tried to follow the initial suggestion on issue #172 e.g. `ConditionalApplyPlugin.withBranchApplyEnabledFor('qa').init()` but I'm new to the terraform-pipeline project and didn't understand the intent of calling `.init()` at the end. I figured I'd take a stab at it and get feedback afterwards.

**Comment by kmanning** [comment]
I wanted to design the library in a way that allowed plugins to be enabled or disabled, as the user sees fit.  I also wanted the enable/disable to be done as simply as possible - a static method fit the bill.  

So `init()` is the convention that I came up with, to enable a plugin - without calling `init()`, a plugin should not have any effect on your pipeline, with the exception of a handful of "default" plugins (which includes ConditionalApplyPlugin) See: https://github.com/manheim/terraform-pipeline/blob/affa2f11c59bde1c388209b681b37d64327835a9/src/TerraformEnvironmentStage.groovy#L7

This is definitely an awkward implementation in hindsight.  A better implementation could have treated ConditionalApplyPlugin (and every default plugin), like any other plugin, and had it implement an `init()` method.  That `init()` method could then have been called in `Jenkinsfile.init()`, and achieved the same current behavior, without the code being any different from any other plugin.

**Comment by kmanning** [comment]
The implementation looks good.  Can you update the CHANGELOG with your Issue?  I'm also about to cut a release on master, and this change can go into v5.14, the next release.

**Comment by kmanning** [comment]
Hrm, I think I spent more time thinking/talking about the `init` method, that I didn't look into the implementation details enough.  I think what you implemented is different from the intent outlined in the issue.

Your change will allow `apply` to run on *all environments* so long as the branch names match.

The intent described in the issue is that `apply` can run on *a specific environment*, regardless of what the branch is named.

Those are two different things.  I pointed you in the direction of the existing Issue - is your UseCase different from what was outlined?  If so, we should open a different issue.

**Comment by kmanning** [comment]
Per the discussion about the awkwardness of `init` and `ConditionalApplyPlugin`, I put in Issue #327

**Comment by kemper** [comment]
After chatting with @kmanning I realize I misinterpreted the goal of this issue. My PR does address half of issue #25 however, so I'm planning to update the PR for that issue.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Unreleased
 
+* [Issue #321](https://github.com/manheim/terraform-pipeline/issues/172) Allow 'apply' on branches other than master for some environments
```
Comment (kmanning) [inline CHANGELOG.md:3]: Can the Issue number and link be corrected?  Both should point to Issue #25 - https://github.com/manheim/terraform-pipeline/issues/25

**Comment by kmanning** [comment]
Other than the links in the ChangeLog, the changes look good.

**Pull request merged**